src/metabase/models/collection.clj

```diff
@@ -630,14 +630,25 @@
                                              (when-not (collection.root/is-root-collection? parent-coll)
                                                [:not= :c2.id (u/the-id parent-coll)])]}]]]))]}]])))
 
-(mu/defn visible-collection-ids :- VisibleCollections
+(def ^{:arglists '([visibility-config])} visible-collection-ids
   "Returns all collection IDs that are visible given the `visibility-config` passed in. (Config provides knobs for
   toggling permission level, trash/archive visibility, etc). If you're trying to filter based on this, you should
-  probably try to use `visible-collection-filter-clause` instead."
-  [visibility-config]
-  (cond-> (t2/select-pks-set :model/Collection {:where (visible-collection-filter-clause :id visibility-config)})
-    (should-display-root-collection? visibility-config)
-    (conj "root")))
+  probably try to use `visible-collection-filter-clause` instead.
+
+  Cached for the lifetime of the request, maximum 10 seconds."
+  (memoize/ttl
+   ^{::memoize/args-fn (fn [[visibility-config]]
+                         (if-let [req-id *request-id*]
+                           [req-id api/*current-user-id* visibility-config]
+                           [(random-uuid) api/*current-user-id* visibility-config]))}
```

Comment:
If there's no request-id, the intent is to not memoize it. Do you think it's worth writing this in a way where we can throw away calls to this where request-id isn't bound?

Reply:
Discussed on slack, takeaway is this is ok for now (mostly because it'd be hard to do otherwise).

Reply:
seems like it would be easy to add a wrapping function here which could call the cached or uncached version depending on the value of `*request-id*`